samples/mixed/provider/src/provider_impl.rs

```diff
@@ -0,0 +1,267 @@
+    ) -> Result<Response<SubscribeResponse>, Status> {
+        let request_inner = request.into_inner();
+        let entity_id: String = request_inner.entity_id.clone();
+        let consumer_uri: String = request_inner.consumer_uri;
+
+        info!(
+            "Received a subscribe request from consumer URI {} for entity id {}",
+            &consumer_uri, &entity_id
+        );
+
+        let mut lock: MutexGuard<HashMap<String, HashSet<String>>> =
+            self.subscription_map.lock().unwrap();
+        let mut uris = match lock.get(&entity_id) {
+            Some(get_value) => get_value.clone(),
+            None => HashSet::new(),
+        };
+        uris.insert(consumer_uri);
+        lock.insert(entity_id, uris);
+
+        // info!("Completed subscription.");
+
+        let response = SubscribeResponse {};
+
+        Ok(Response::new(response))
+    }
+
+    /// Unsubscribe implementation.
+    ///
+    /// # Arguments
+    /// * `request` - Unsubscribe request.
+    async fn unsubscribe(
+        &self,
+        request: Request<UnsubscribeRequest>,
+    ) -> Result<Response<UnsubscribeResponse>, Status> {
+        warn!("Got an unsubscribe request: {request:?}");
+
+        Err(Status::unimplemented("unsubscribe has not been implemented"))
+    }
+
+    /// Get implementation.
+    ///
+    /// # Arguments
+    /// * `request` - Get request.
+    async fn get(&self, request: Request<GetRequest>) -> Result<Response<GetResponse>, Status> {
+        warn!("Got a get request: {request:?}");
+
+        Err(Status::unimplemented("get has not been implemented"))
+    }
+
+    /// Set implementation.
+    ///
+    /// # Arguments
+    /// * `request` - Set request.
+    async fn set(&self, request: Request<SetRequest>) -> Result<Response<SetResponse>, Status> {
+        warn!("Got a set request: {request:?}");
+
+        Err(Status::unimplemented("set has not been implemented"))
+    }
+
+    /// Invoke implementation.
+    ///
+    /// # Arguments
+    /// * `request` - Invoke request.
+    async fn invoke(
+        &self,
+        request: Request<InvokeRequest>,
+    ) -> Result<Response<InvokeResponse>, Status> {
+        debug!("Got an invoke request: {request:?}");
+
+        let request_inner = request.into_inner();
+        let entity_id: String = request_inner.entity_id.clone();
+        let response_id: String = request_inner.response_id.clone();
+        let consumer_uri: String = request_inner.consumer_uri;
+        let payload: String = request_inner.payload;
+
+        debug!(
+            "Received an invoke request from consumer URI {} for entity id {} with payload '{}'",
+            &consumer_uri, &entity_id, &payload
+        );
+
+        let vehicle: Arc<Mutex<Vehicle>> = self.vehicle.clone();
+
+        tokio::spawn(async move {
+            let mut response_payload: String = format!("Successfully invoked {entity_id}");
+
+            if entity_id == ACTIVATE_AIR_CONDITIOING {
```

Comment:
https://doc.rust-lang.org/rust-by-example/flow_control/match.html ?

Reply:
It has a compile time error, as match does not work with string slices.

Also, this is a personal preference thing.  I'm leaving it as is.